exchanges/ftx/ftx_wrapper.go

```diff
@@ -1254,3 +1266,171 @@ func (f *FTX) GetAvailableTransferChains(ctx context.Context, cryptocurrency cur
 	}
 	return availableChains, nil
 }
+
+// CalculatePNL determines the PNL of a given position based on the PNLCalculatorRequest
+func (f *FTX) CalculatePNL(ctx context.Context, pnl *order.PNLCalculatorRequest) (*order.PNLResult, error) {
+	if pnl == nil {
+		return nil, fmt.Errorf("%v %w", f.Name, order.ErrNilPNLCalculator)
+	}
+	result := &order.PNLResult{
+		Time: pnl.Time,
+	}
+	var err error
+	if pnl.CalculateOffline {
+		// PNLCalculator matches FTX's pnl calculation method
+		calc := order.PNLCalculator{}
+		result, err = calc.CalculatePNL(ctx, pnl)
+		if err != nil {
+			return nil, fmt.Errorf("%s %s %w", f.Name, f.API.Credentials.Subaccount, err)
+		}
+	}
+
+	ep := pnl.EntryPrice.InexactFloat64()
+	info, err := f.GetAccountInfo(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if info.Liquidating || info.Collateral == 0 {
+		result.IsLiquidated = true
+		return result, fmt.Errorf("%s %s %w", f.Name, f.API.Credentials.Subaccount, order.ErrPositionLiquidated)
+	}
+	for i := range info.Positions {
+		var pair currency.Pair
+		pair, err = currency.NewPairFromString(info.Positions[i].Future)
+		if err != nil {
+			return nil, err
+		}
+		if !pnl.Pair.Equal(pair) {
+			continue
+		}
+		if info.Positions[i].EntryPrice != ep {
+			continue
+		}
+		result.UnrealisedPNL = decimal.NewFromFloat(info.Positions[i].UnrealizedPNL)
+		result.RealisedPNLBeforeFees = decimal.NewFromFloat(info.Positions[i].RealizedPNL)
+		result.Price = decimal.NewFromFloat(info.Positions[i].Cost)
+		return result, nil
+	}
+	// order no longer active, use offline calculation
+	calc := order.PNLCalculator{}
+	result, err = calc.CalculatePNL(ctx, pnl)
+	if err != nil {
+		return nil, fmt.Errorf("%s %s %w", f.Name, f.API.Credentials.Subaccount, err)
+	}
+	return result, nil
+}
+
+// ScaleCollateral takes your totals and scales them according to FTX's rules
+func (f *FTX) ScaleCollateral(ctx context.Context, calc *order.CollateralCalculator) (decimal.Decimal, error) {
+	var result decimal.Decimal
+	if calc.CalculateOffline {
+		if calc.CollateralCurrency.Match(currency.USD) {
+			// FTX bases scales all collateral into USD amounts
+			return calc.CollateralAmount, nil
+		}
+		collateralWeight, ok := f.collateralWeight[calc.CollateralCurrency.Upper().String()]
+		if !ok {
+			return decimal.Zero, fmt.Errorf("%s %s %w", f.Name, calc.CollateralCurrency, errCollateralCurrencyNotFound)
+		}
+		if calc.CollateralAmount.IsPositive() {
```

Comment:
NOTE: This is not a change request but we might be able to get a better value with something like this included. 
```go
if calc.USDPrice.Equal(decimal.Zero) {
			last, err := ticker.FindLast(currency.NewPair(calc.CollateralCurrency, currency.USD), calc.Asset)
			if err != nil {
				log.Warn(log.Global, "not found bro")
			}
			calc.USDPrice = decimal.NewFromFloat(last)
		}

```

Also when doing online I am coming across zero value for SRM_LOCKED which is expected but other tokens which are zero.

Reply:
This was a good bug catch. Offline calculation wasn't sending USD values. I think I'll put it at the rcpserver level. 

Does "offline calculation" as a command term seem accurate to you? By offline, I mean that when scaling collateral, it already has all the data it needs (which will be used in the backtester for example), but for this RPCServer function, I'll need to retreive USD value data per currency.

Also, what are you asking of me for the online part?